### PR TITLE
Fix: issue#190

### PR DESCRIPTION
**Pull request by hajime-nagisa**
I really wanted to use this awesome plugin, but I couldn't do so because the issue #190 is critical for Japanese. "バカ", "ハカ" and "パーカー" have completely different meaning, but these characters are incorrectly rendered as "ハカ".

So I tried to fix it.

Things I confirmed:
 - The ignore diacritics setting is working correctly
 - Characters with diacritical mark are displayed correctly regardless of ignore diacritics setting
 - Highlighting works correctly with alphabets, Cyrillic alphabets, Japanese Hiragana/Katakana, Kanji and Hangul even if when enclosed in quotation marks

I'm not confident about my implementations as a beginner developer. Especially, I have concern about the changes in caching and the behavior of `getMatches()` . The use of string index might be prone to bugs, but I couldn't come up with other ideas...

I would greatly appreciate a thorough review. Thank you!

### CODE REVIEW

**Comment by scambier** [comment]
Thanks for your PR, I'll review it asap :)

**Comment by scambier** [comment]
Given that Omnisearch is messier than I'd like, I think you did a great job at understanding its workflow, and your changes look good.

> I have concern about (...) the behavior of getMatches() . The use of string index might be prone to bugs, but I couldn't come up with other ideas...

Using string indexes is, I think, the best solution to solve this issue. But there's indeed a bug, probably in `stringsToRegex()` and/or `getMatches()` :)

To reproduce:
- Create a note with the following contents
```
test109.foo
test109,bar
```
- Search for "test109". The note appears in the results, but no highlighting.
- Modify the note with the following contents. Note the space after the comma on the second line
```
test109.foo
test109, bar
```
- Search for "test109". The highlight works as expected on both lines

In this other example, "llkkjj" is only hightlighted when "test" isn't present in the search field
image
image

For reference, here's the current behavior (master branch):

image
image
image

Not perfect, but it at least highlights the correct "word".

---

Excellent job overall, I'll gladly merge your PR as soon as this bug is fixed 👍

_(and I should definitely write unit tests for text-processing.ts...)_

**Comment by hajime-nagisa** [comment]
Thank you for the review! I'm reassured to know that my code isn't so bad. 

I've fixed the bug you mentioned. Now, it correctly highlights the word inputted in the query. However, I noticed it doesn't highlight commas and periods. I tried to fix this, but it seems to require modifications to codes about Minisearch. It seems to be quite challenging, so I gave up... I think it won't significantly affect the UX.

Here are the screenshots.
スクリーンショット 2023-11-26 11 15 37
スクリーンショット 2023-11-26 11 15 16
スクリーンショット 2023-11-26 11 14 54
スクリーンショット 2023-11-26 11 15 03

I hope there are no more bugs in the code. I'm looking forward to its integration!

**Comment by scambier** [comment]
I tested the first token separators (list at the bottom of globals.ts), `^` doesn't work either, and most others like `? % / ;` (not exhaustive) are highlighted but not split:
image

Ideally the `separators` list should be reused _as is_ for highlighting (instead of specific instances like `/[.,]/`) to make sure that all (future) cases are correctly covered 👍

> However, I noticed it doesn't highlight commas and periods.

The commas highlights was a side-effect due to the word separator `\b`. I used `\b` because it was overall simpler and cheaper to do the highlights this way. It's not expected to highlight token separators since they're not indexed.

**Comment by hajime-nagisa** [comment]
Thank you for the testing. I've fixed the code following your advice. I haven't tested all the separators, but I confirmed it working with the separators I usually use.

**Review by scambier — APPROVED** [review]

File: src/tools/utils.ts [hunk 1]
```diff
@@ -126,9 +126,12 @@ export function removeDiacritics(str: string): string {
   // Keep backticks for code blocks, because otherwise they are removed by the .normalize() function
   // https://stackoverflow.com/a/36100275
   str = str.replaceAll('`', '[__omnisearch__backtick__]')
+  // Keep caret same as above
+  str = str.replaceAll('^', '[__omnisearch__caret__]')
```
[the remark this answers is not in the document]
Comment (hajime-nagisa) [reply src/tools/utils.ts:130]: Sorry for the late response. Without the code you mentioned, the text won't be highlighted if the query includes a caret between words. This is due to the normalize function, which removes diacritics from the text, thereby causing the string's index to become incorrect.

Here are the screenshots:
スクリーンショット 2023-11-29 16 30 46
スクリーンショット 2023-11-29 16 30 54

**Comment by hajime-nagisa** [comment]
I'm so glad the pull request was merged. It was a very good experience for me!
Thank you for your support overall🙏

**Comment by scambier** [comment]
And thanks for your great work :)

I'll publish a beta asap 👍

**Pull request merged**